src/audio/xm64.c

```diff
@@ -0,0 +1,240 @@
+			mixer_ch_play(first_ch+i, w);
+			mixer_ch_set_pos(first_ch+i, ch->sample_position);
+
+			// Configure also frequency and volume that might have changed
+			// since last tick.
+			// debugf("vol %f %f\n", ch->actual_volume, ch->actual_panning);
+			mixer_ch_set_freq(first_ch+i, ch->frequency);
+			mixer_ch_set_vol(first_ch+i,
+				muted ? 0 : gvol * ch->actual_volume[0],
+				muted ? 0 : gvol * ch->actual_volume[1]);
+		} else {
+			// No sample in this channel: the channel is mute. Just stop it.
+			mixer_ch_stop(first_ch+i);
+		}
+	}
+
+	// Schedule next tick according to the number of samples in this tick.
+	int delay = ceilf(ctx->remaining_samples_in_tick);
+	ctx->remaining_samples_in_tick -= delay;
+	return delay;
+}
+
+void xm64player_open(xm64player_t *player, const char *fn) {
+	memset(player, 0, sizeof(*player));
+
+	// No pending seek at the moment, we start from beginning anyway.
+	player->seek.patidx = -1;
+
+	char romfn[strlen(fn)+5+1];
+	snprintf(romfn, sizeof(romfn), "rom:/%s", fn);
+	player->fh = fopen(romfn, "rb");
+	assertf(player->fh, "Cannot open file: %s", fn);
+
+	// Load the XM context
+	int sample_rate = audio_get_frequency();
+	assertf(sample_rate >= 0, "audio_init() and mixer_init() must be called before xm64player_open()");
+	if (xm_context_load(&player->ctx, player->fh, sample_rate) != 0) {
+		// Check if the file looks like a standard XM, so to provide
+		// a clear message in that case.
+		char signature[16] = {0};
+		fseek(player->fh, 0, SEEK_SET);
+		fread(signature, 1, 15, player->fh);
+		if (strcmp(signature, "Extended Module") == 0) {
+			assertf(0, "cannot load XM file %s -- please convert to XM64 with audio64", fn);
+		}
+		assertf(0, "error loading XM64 file -- file corrupted?");
+	}
+
+	uint32_t base_rom_addr = dfs_rom_addr(fn);
+
+	// Count samples
+	int ninst = xm_get_number_of_instruments(player->ctx);
+	int nwaves = 0;
+	for (int i=0;i<ninst;i++)
+		nwaves += xm_get_number_of_samples(player->ctx, i+1);
+
+	// Allocate waveforms (one per XM64's "samples" aka waveforms)
+	player->waves = malloc(sizeof(waveform_t) * nwaves);
+	assert(player->waves);
+	int nw = 0;
+	for (int i=0;i<ninst;i++) {
+		xm_instrument_t *inst = &player->ctx->module.instruments[i];
+		for (int j=0;j<inst->num_samples;j++) {
+			xm_sample_t *samp = &inst->samples[j];
+
+			// Convert offset of samples from relative to absolute
+			samp->data8_offset += base_rom_addr;
+
+			// Initialize the waveform_t structures with information
+			// coming from the XM "sample".
+			samp->wave = &player->waves[nw++];
+			memset(samp->wave, 0, sizeof(waveform_t));
+			samp->wave->name = strdup(fn); // FIXME: maybe better use a proper name here
+			samp->wave->bits = samp->bits;
+			samp->wave->channels = 1;
+			samp->wave->frequency = sample_rate; // fake, will be changed at every key-on
+			samp->wave->len = samp->length;
+			samp->wave->loop_len = samp->loop_type == XM_NO_LOOP ? 0 : samp->loop_length;
+			if (samp->wave->loop_len&1)
```

Comment:
Right, there's no comment. This is explained in `wav64_set_loop`. This is the comment there:

```
	// Odd loop lengths are not supported for 8-bit waveforms because they would
	// change the 2-byte phase between ROM and RDRAM addresses during loop unrolling.
	// We shorten the loop by 1 sample which shouldn't matter.
	// Notice that audioconv64 does the same during conversion.
```

The only real constraint for `dma_read` now is that the RDRAM and ROM addresses must have the same 2-byte phase (that is, they must be both odd or both even). The reason is that physical DMA is actually 16-bit aligned so you can only run it on even addresses, but if you have two odd addresses, dma_read will just move one byte manually before starting the transfer.

In case of audio, in general you need to fetch samples from ROM into the RDRAM samplebuffer provided by the mixer. Normally they start aligned (both ROM waveforms and RDRAM samplesbufers starts as 8-byte aligned), and you always keep this alignment as you go. When the samplebuffer wraps around, the code in samplebuffer_discard (used as part of looping around in the buffer) will make sure not to lose the 2-byte phase. So in general the phase is always kept. The only exception would be a 8-bit waveform that declares a loop of odd size. In that case, the last byte of the waveform will be correctly phases, but the following byte (beginning of the loop) would be out of phase. We don't have a solution for this, we currently do not support this scenario, and we simply shorten the loop by 1 sample.

The code you've highlighted is technically wrong because it does the shortening also for 16-bit waveforms where it is not required. I'll fix it a document it, thanks.